### PR TITLE
Upgrades overhaul part 4b: UpgradeActorsNear -> ProximityExternalCondition

### PR DESCRIPTION
**Pull request by pchote**
Depends on #12406.

### CODE REVIEW

**Comment by pchote** [comment]
Rebased and ready for review.

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeActorsNear.cs [hunk 1]
```diff
@@ -9,16 +9,17 @@
...
 #endregion
 
+using System.Collections.Generic;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Applies an upgrade to actors within a specified range.")]
-	public class UpgradeActorsNearInfo : ITraitInfo
+	public class ProximityExternalConditionInfo : ITraitInfo
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeActorsNear.cs]: Why didn't you rename the file as well?

File: OpenRA.Mods.Common/Traits/Upgrades/UpgradeActorsNear.cs [hunk 2]
```diff
@@ -129,26 +131,24 @@ public void UnitProducedByOther(Actor self, Actor producer, Actor produced)
 					return;
 
 				var um = produced.TraitOrDefault<UpgradeManager>();
-				if (um != null)
-					foreach (var u in info.Upgrades)
-						if (um.AcknowledgesUpgrade(produced, u))
-							um.GrantTimedUpgrade(produced, u, 1);
+				if (um != null && um.AcceptsExternalCondition(produced, info.Condition))
+					tokens[produced] = um.GrantCondition(produced, info.Condition);
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Upgrades/UpgradeActorsNear.cs]: Shouldn't you also set the `external` flag here?

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 3]
```diff
@@ -555,6 +555,15 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				if (engineVersion < 20161202)
```
Comment (obrakmann) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Don't forget to update this eventually.

File: mods/ra/maps/allies-03a/rules.yaml [hunk 4]
```diff
@@ -34,6 +34,8 @@ HACKE6:
 		CaptureTypes: building
 	Targetable:
 		RequiresCondition: !jail
+	ExternalConditions:
+		Conditions: jail
```
Comment (obrakmann) [inline mods/ra/maps/allies-03a/rules.yaml]: You need to do this for the medics as well here, and for both medics and hacke6 in 3b.

**Review by obrakmann — CHANGES_REQUESTED** [review]

**Comment by pchote** [comment]
Fixed and rebased.

**Review by obrakmann — APPROVED** [review]

File: OpenRA.Mods.Common/Traits/Upgrades/ProximityExternalCondition.cs [hunk 5]
```diff
@@ -129,26 +131,24 @@ public void UnitProducedByOther(Actor self, Actor producer, Actor produced)
 					return;
 
 				var um = produced.TraitOrDefault<UpgradeManager>();
-				if (um != null)
-					foreach (var u in info.Upgrades)
-						if (um.AcknowledgesUpgrade(produced, u))
-							um.GrantTimedUpgrade(produced, u, 1);
+				if (um != null && um.AcceptsExternalCondition(produced, info.Condition))
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Upgrades/ProximityExternalCondition.cs]: I assume this is intentional, but I'm struggling to get my head around why this doesn't need a `ContainsKey` check, could you elaborate a bit?

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Upgrades/ProximityExternalCondition.cs]: Okay, thanks.

**Comment by reaperrr** [comment]
👍

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f2bc3962fcf89a42e0989239f847006b46accafa...6894f77f04517a6684ad581fdea1a8fbe92fe7cf)

**Pull request merged**